### PR TITLE
feat: prisma application flagged sets

### PR DESCRIPTION
**Pull request by YazeedLoonat**
# Pull Request Template

## Issue Overview

This PR addresses [#issue](https://github.com/bloom-housing/bloom/issues/3540)

- [x] This change addresses the issue in full

## Description
This creates the services and end points for application flagged sets.

**Part of the ask on this one is to rework our current process flow**
We have been having issues where the current (typeorm version) of the process endpoint/cronjob is causing issues for existing flagged sets. This PR reworks that flow entirely to avoid that problem. That's probably the part of this pr I think needs the heaviest review to make sure no edge cases snuck through

## How Can This Be Tested/Reviewed?
Unit and Integration tests were added, but you can also use the openapi site at http://localhost:3101/api/ to test further

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [x] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [x] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

File: backend_new/test/unit/services/application-flagged-set.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,2431 @@
...
+    });
+  });
+
+  it('should grab meta data', async () => {
+    const mockCount = jest.fn();
+    mockCount.mockResolvedValueOnce(1);
+    mockCount.mockResolvedValueOnce(2);
+    mockCount.mockResolvedValueOnce(3);
+    mockCount.mockResolvedValueOnce(4);
+    mockCount.mockResolvedValueOnce(5);
```
Comment (ludtkemorgan) [inline backend_new/test/unit/services/application-flagged-set.service.spec.ts:529]: I don't know if it is any cleaner, but jest.fn() is a chainable call so this could be 
```
    const mockCount = jest
      .fn()
      .mockResolvedValueOnce(1)
      .mockResolvedValueOnce(2)
      .mockResolvedValueOnce(3)
      .mockResolvedValueOnce(4)
      .mockResolvedValueOnce(5);

File: backend_new/src/services/application-flagged-set.service.ts [hunk 2]
```diff
@@ -0,0 +1,951 @@
...
+      totalPendingCount,
+      totalNamePendingCount,
+      totalEmailPendingCount,
+    };
+  }
+
+  /**
+    helper that builds the meta functions queries
+  */
+  metaHelper(
```
Comment (ludtkemorgan) [inline backend_new/src/services/application-flagged-set.service.ts:248]: I wonder if renaming this to something more applicable about getting counts would make sense. I'm not great with names, but maybe something like "afsCountQuery"

**Review by ludtkemorgan — COMMENTED** [review]
No big issues have come up yet. I still need to do some functional testing though

**Comment by YazeedLoonat** [comment]
From convo with Morgan, add back in the process on listing update (but do it only for that specific listing instead of for every listing)

**Comment by ludtkemorgan** [comment]
@YazeedLoonat another thing I thought of was it would be nice to have some of the seeded data be applications that get flagged. That doesn't need to be part of this PR though

File: backend_new/src/dtos/application-flagged-sets/afs-query-params.dto.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  listingId: string;
+
+  @Expose()
+  @ApiPropertyOptional({
+    enum: View,
+    enumName: 'AfsView',
+    example: 'active',
+  })
+  @IsEnum(View, { groups: [ValidationsGroupsEnum.default] })
+  view?: View;
```
Comment (ColinBuyck) [inline backend_new/src/dtos/application-flagged-sets/afs-query-params.dto.ts:21]: Surprised to see the afs get endpoint fail (Bad Request) without a view selected even though its marked as optional here
Screenshot 2023-10-13 at 11 46 35 AM

File: backend_new/.env.template [hunk 4]
```diff
@@ -34,4 +34,6 @@ TWILIO_AUTH_TOKEN=
 PARTNERS_PORTAL_URL=http://localhost:3001/
 # sendgrid email api key
 EMAIL_API_KEY=SG.ExampleApiKey
+# controls the repititon of the afs cron job
```
Comment (emilyjablonski) [inline backend_new/.env.template:37]: `repetition`

File: backend_new/src/dtos/application-flagged-sets/afs-meta.dto.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  @IsNumber()
+  @IsOptional()
+  @ApiPropertyOptional()
+  totalResolvedCount?: number;
+
+  @Expose()
+  @IsNumber()
+  @IsOptional()
+  @ApiPropertyOptional()
+  totalPendingCount?: number;
```
Comment (emilyjablonski) [inline backend_new/src/dtos/application-flagged-sets/afs-meta.dto.ts:22]: can / should this be a computed value based on the two below? somehow seems redundant

File: backend_new/src/services/application-flagged-set.service.ts [hunk 6]
```diff
@@ -0,0 +1,951 @@
...
+import { OrderByEnum } from '../enums/shared/order-by-enum';
+import { View } from '../enums/application-flagged-sets/view';
+import { buildPaginationMetaInfo } from '../utilities/pagination-helpers';
+import { AfsResolve } from '../dtos/application-flagged-sets/afs-resolve.dto';
+import { User } from '../dtos/users/user.dto';
+import { Application } from '../dtos/applications/application.dto';
+import { IdDTO } from 'src/dtos/shared/id.dto';
+
+/*
+  this is the service for unit types
```
Comment (emilyjablonski) [inline backend_new/src/services/application-flagged-set.service.ts:36]: is this comment accurate?

File: backend_new/src/services/application-flagged-set.service.ts [hunk 7]
```diff
@@ -0,0 +1,951 @@
...
+      application.applicant.birthDay,
+      ...(application.householdMember
+        ? application.householdMember.map((member) => member.birthDay)
+        : []),
+    ];
+    const birthYears = [
+      application.applicant.birthYear,
+      ...(application.householdMember
+        ? application.householdMember.map((member) => member.birthYear)
+        : []),
```
Comment (seanmalbert) [inline backend_new/src/services/application-flagged-set.service.ts:727]: suggestion (non-blocking): you could create a function to handle each of these that accepted a "key" parameter like "lastName" and returned the array.

File: backend_new/src/services/application-flagged-set.service.ts [hunk 8]
```diff
@@ -0,0 +1,951 @@
...
+        },
+      });
+    }
+  }
+
+  /**
+    tests application to see if its a duplicate
+    if it is then its either added to a flagged set or a new flagged set is created
+  */
+  async testApplication(
```
Comment (seanmalbert) [inline backend_new/src/services/application-flagged-set.service.ts:547]: question: I think this all checks out, but could you spend some time thinking about how this function could be made more readable?

**Review by seanmalbert — CHANGES_REQUESTED** [review]

File: backend_new/.env.template [hunk 9]
```diff
@@ -34,4 +34,6 @@ TWILIO_AUTH_TOKEN=
...
+AFS_PROCESSING_CRON_STRING=0,3,15 * * * *
```
Comment (seanmalbert) [inline backend_new/.env.template:38]: question: for clarification this means it will run at minute 0, 3 and 15, correct?

suggestion: if my understanding is correct, recommend having cron run at more even intervals like "0, 15, 30, 45" or something like "15, 45", or just once an hour. Not sure what the optimal frequency will be that won't cause performance issues under heavy load.

File: backend_new/package.json [hunk 10]
```diff
@@ -38,12 +38,14 @@
...
     "@nestjs/platform-express": "^8.0.0",
     "@nestjs/swagger": "~7.1.12",
+    "@nestjs/schedule": "^3.0.4",
     "@prisma/client": "^5.0.0",
     "@sendgrid/mail": "7.7.0",
     "class-transformer": "~0.5.1",
     "class-validator": "~0.14.0",
     "cloudinary": "^1.37.3",
     "cookie-parser": "~1.4.6",
+    "cron": "^2.4.3",
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/package.json:48]: removed!

File: backend_new/prisma/seed-helpers/application-factory.ts [hunk 11]
```diff
@@ -16,6 +16,7 @@ export const applicationFactory = (optionalParams?: {
   applicant?: Prisma.ApplicantCreateWithoutApplicationsInput;
   overrides?: Prisma.ApplicationsCreateInput;
   listingId?: string;
+  householdMemeber?: Prisma.HouseholdMemberCreateWithoutApplicationsInput;
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/prisma/seed-helpers/application-factory.ts:19]: hehe fixed

File: backend_new/test/integration/ami-chart.e2e-spec.ts [hunk 12]
```diff
@@ -25,6 +26,11 @@ describe('AmiChart Controller Tests', () => {
     prisma = moduleFixture.get<PrismaService>(PrismaService);
     await app.init();
 
+    const schedulerRegistry =
+      moduleFixture.get<SchedulerRegistry>(SchedulerRegistry);
+    // we stop the cron job since we don't want the cron job to run during tests
+    schedulerRegistry.getCronJob('AFS_CRON_JOB').stop();
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/test/integration/ami-chart.e2e-spec.ts:32]: ooo that's such a good idea! fixed

File: backend_new/test/integration/application-flagged-set.e2e-spec.ts [hunk 13]
```diff
@@ -0,0 +1,1521 @@
...
+        applications: true,
+      },
+    });
+
+    expect(afs.length).toEqual(1);
+    expect(afs[0].rule).toEqual(RuleEnum.email);
+
+    await prisma.applications.update({
+      data: {
+        acceptedTerms: true,
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/test/integration/application-flagged-set.e2e-spec.ts:723]: yeah this is just a field to update so the update date gets auto updated

File: backend_new/test/integration/application-flagged-set.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,1521 @@
...
+        allApplicationCount++;
+      } else if (
+        flaggedSet.ruleKey.indexOf(`${listing}-email3@email.com`) >= 0
+      ) {
+        const applications = flaggedSet.applications.map((app) => app.id);
+        expect(applications).toContain(appC.id);
+        allApplicationCount++;
+        expect(applications).toContain(appD.id);
+        allApplicationCount++;
+      }
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/test/integration/application-flagged-set.e2e-spec.ts:1166]: updated!

File: backend_new/test/integration/application-flagged-set.e2e-spec.ts [hunk 15]
```diff
@@ -0,0 +1,1521 @@
...
+      },
+      include: {
+        applications: true,
+      },
+    });
+
+    expect(afs.length).toEqual(1);
+    expect(afs[0].rule).toEqual(RuleEnum.email);
+  });
+});
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/test/integration/application-flagged-set.e2e-spec.ts:1667]: added!

File: backend_new/test/unit/services/application-flagged-set.service.spec.ts (hunk 1 above)
Comment (YazeedLoonat) [reply backend_new/test/unit/services/application-flagged-set.service.spec.ts]: updated!

File: backend_new/src/services/application-flagged-set.service.ts (hunk 2 above)
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts]: I rolled with `metaDataQueryBuilder` hopefully thats more descriptive

File: backend_new/src/dtos/application-flagged-sets/afs-query-params.dto.ts (hunk 3 above)
Comment (YazeedLoonat) [reply backend_new/src/dtos/application-flagged-sets/afs-query-params.dto.ts:21]: So view in this case isn't always passed from the front end, so I think making it required would be kinda bad 

I'm not sure what change I made that seems to have fixed this, but it seems like its fixed from what I can tell

File: backend_new/.env.template (hunk 4 above)
Comment (YazeedLoonat) [reply backend_new/.env.template]: ty!

File: backend_new/src/dtos/application-flagged-sets/afs-meta.dto.ts (hunk 5 above)
Comment (YazeedLoonat) [reply backend_new/src/dtos/application-flagged-sets/afs-meta.dto.ts:22]: I think we could for sure calculate it that way, but if we need this data in more than just the partner site (like for a future reporting tool or something) its probably for the best we just calculate that on the backend and let the front end know what the value is 

hopefully that makes some sense

File: backend_new/src/services/application-flagged-set.service.ts (hunk 6 above)
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts]: my secrets have been revealed copy and pasting from other services lol

fixed!

File: backend_new/src/services/application-flagged-set.service.ts [hunk 16]
```diff
@@ -0,0 +1,951 @@
...
+          where: {
+            name: CRON_JOB_NAME,
+          },
+        });
+        // To prevent multiple overlapped jobs only run if one hasn't started in the last 5 minutes
+        if (
+          !currentCronJob ||
+          currentCronJob.lastRunDate <
+            dayjs(new Date()).subtract(5, 'minutes').toDate()
+        ) {
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts]: from convo name has been made unique in the db so collisions should be less likely/impossible

File: backend_new/src/services/application-flagged-set.service.ts [hunk 17]
```diff
@@ -0,0 +1,951 @@
...
+        filters.push({
+          status: FlaggedSetStatusEnum.pending,
+          rule: RuleEnum.email,
+        });
+      } else if (params.view === View.resolved) {
+        filters.push({
+          status: FlaggedSetStatusEnum.resolved,
+        });
+      }
+    }
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts]: updated!

File: backend_new/src/services/application-flagged-set.service.ts [hunk 18]
```diff
@@ -0,0 +1,951 @@
...
+        params.listingId,
+        FlaggedSetStatusEnum.pending,
+        RuleEnum.nameAndDOB,
+      ),
+      this.metaHelper(
+        params.listingId,
+        FlaggedSetStatusEnum.pending,
+        RuleEnum.email,
+      ),
+    ]);
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts]: 2 less queries! updated 

I think the backend should still be responsible for the calculation so I kept it as part of the dto, we just no do the calculation on the fly instead of reading from the db again!

File: backend_new/src/services/application-flagged-set.service.ts [hunk 19]
```diff
@@ -0,0 +1,951 @@
...
+
+    if (afs.listings.status !== ListingsStatusEnum.closed) {
+      throw new BadRequestException(
+        `Listing ${afs.listings.id} must be closed before resolving any duplicates`,
+      );
+    }
+
+    const selectedApps = dto.applications?.length
+      ? afs.applications.map((app) => app.id)
+      : [];
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts]: fixed, good call out

File: backend_new/src/services/application-flagged-set.service.ts [hunk 20]
```diff
@@ -0,0 +1,951 @@
...
+    const afs = await this.prisma.applicationFlaggedSet.findFirst({
+      where: {
+        AND: filter,
+      },
+      include: {
+        listings: true,
+        applications: {
+          where: {
+            id: {
+              in: dto.applications.map((application) => application.id),
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts]: updated!

[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts]: yeah so its kinda an interesting thing that prisma does
the one in the where clause would only get flagged sets where the application is attached (applications.some.id.i)

the other one (include.applications.where.id.in) is so that when we get the list of application flagged sets back the applications that are returned as part of that call are only those that meet the criteria

without that second one we would just be getting the entire set of applications on the flagged set. Hopefully that makes some sense 

https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#filter-a-list-of-relations

is the documentation for it

File: backend_new/src/services/application-flagged-set.service.ts (hunk 7 above)
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts]: nice call out! updated

File: backend_new/src/services/application-flagged-set.service.ts (hunk 8 above)
Comment (YazeedLoonat) [reply backend_new/src/services/application-flagged-set.service.ts:562]: Hey sean, I'm not really sure how to condense or pull this function apart further. I thought about making some more helpers but I feel like it might make this file/function a bit "spaghetti-y" with having to jump around to figure out whats its trying to do 

I made some small tweaks, but lemme know if you have anything in mind to really clean this up

File: backend_new/.env.template (hunk 9 above)
Comment (YazeedLoonat) [reply backend_new/.env.template:38]: That's a really good call out, on production we use `0 0 * * *` as the value 
which we mean every day at midnight server time this would execute 
we add the seconds as a random number 

Right now we have this run so frequently locally so that we can catch any issues with the process job, I can change the suggestion in the .env.template though if that would make things more clear

**Comment by YazeedLoonat** [comment]
Hey @seanmalbert for: https://github.com/bloom-housing/bloom/pull/3674#discussion_r1370854777

I think we need to keep the view enum as optional. Since the front end makes requests to endpoints without a view in a couple places. I did fix the example though

**Comment by YazeedLoonat** [comment]
Hey @ludtkemorgan @seanmalbert @emilyjablonski @ColinBuyck 
this should be ready for 👀 again!

**Review by ludtkemorgan — APPROVED** [review]

File: backend_new/.env.template (hunk 9 above)
Comment (seanmalbert) [reply backend_new/.env.template:38]: ah, thanks for the clarification. I think it's ok to keep what's in the template, but maybe add to the comment what is recommended for production environments.

File: backend_new/src/services/application-flagged-set.service.ts (hunk 8 above)
Comment (seanmalbert) [reply backend_new/src/services/application-flagged-set.service.ts:562]: Thanks, @YazeedLoonat ,I'm not sure I do without making it harder to follow in other ways like you're saying (at least not without spending more time).

File: backend_new/src/services/application-flagged-set.service.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (seanmalbert) [reply backend_new/src/services/application-flagged-set.service.ts]: Thanks, @YazeedLoonat . Agreed on keeping it part of the DTO.

**Review by seanmalbert — APPROVED** [review]
